// arma3-rcon/app.js
/*
  → app.js
  → arma3-rcon by dotFionn
*/

const be = require('battle-node');

be.prototype.close = function (cb = () => {}) {
  try {
    this.connected = false;
    clearInterval(this.keepalive);
    this.socket.unref();
    this.socket.close();
    cb();
  } catch (e) {
    console.log(e);
  }
};

function Arma3Rcon(ip, port, password, autoReconnectOptions = {}, tokenId = null, serverId = null, req = null, res = null) {
  this.ip = ip;
  this.port = port;
  this.password = password;
  this.tokenId = tokenId;
  this.serverId = serverId;
  this.req = req;
  this.res = res;

  this.autoReconnect = autoReconnectOptions.enabled ?? true;
  this.autoReconnectInterval = autoReconnectOptions.interval ?? 5;
  this.autoReconnectCount = autoReconnectOptions.count ?? 24;

  this.be;

  this.connectCallback = () => {};

  this.connect = function () {
    return new Promise((res, rej) => {
      this.be = new be({ ip: this.ip, port: this.port, rconPassword: this.password });

      this.be.on('login', (err, success) => {
        if (err) {
          return rej(err);
        }

        res(success);

        this.setupEventHandlers();
      });

      this.be.login();
    });
  }.bind(this);

  this.close = function () {
    return new Promise((res, rej) => {
      try {
        if (this.be) {
          this.be.close(res);
        }
        this.be = undefined;
      } catch (e) {
        rej(e);
      }
    });
  }.bind(this);

  this.setupEventHandlers = function () {
    if (!this.be) return;

    this.be.on(
      'disconnected',
      function disconnectHandler() {
        console.log(`RCON disonnected (${this.ip}:${this.port})`);

        if (this.autoReconnect) {
          let attemptsRemaining = this.autoReconnectCount;
          console.log(`reconnecting RCON...`);

          let t = setInterval(
            function () {
              console.log('attempting reconnect...');
              this.connect()
                .then(function (success) {
                  console.log('connected');
                  if (success || attemptsRemaining == 0) {
                    clearInterval(t);
                  }
                })
                .catch(function () {});

              attemptsRemaining--;
            }.bind(this),
            this.autoReconnectInterval * 1000
          );
        }
      }.bind(this)
    );

    this.be.on('message', (c) => {
      const utils = require(`${global["homedir"]}\\utils.js`)
      let _data = { serverId: this.serverId, tokenId: this.tokenId, messages: c, time: new Date().toISOString() }
      
      if (global["config"]["rconlog"])
      {
          console.log(`\x1b[33m[+] ${_data}`)
      }
      utils.insertNewServerWebControlRconChat(_data).then(_ => {}).catch(err => {console.log(err);})

      for (var ak in global["config"]["rcon"][req._user["tokenId"]]["users"]){
        let user = global["config"]["rcon"][req._user["tokenId"]]["users"][ak];

        this.req["subscribeData"] = 
        {
          rcon: 1,
          serverId: user.serverId, 
          authKey: user.authKey, 
          messages: c,
          time: new Date().toISOString()
        };

        // monitor chat and look for keywords

        
        const keywords = ["!test", "!onlineplayers", "!am I cool?"]

        // words that are not allowed and warn players.. 
        //if the warn cound is more then 2 then he gets banned for a day lol
        const notAllowedWords = [];
        
        keywords.forEach(element => {
          let match = new RegExp(element, 'g');
          let found = match.test(c);

          if (found) {
            console.log("found a keyword in the message... " + element + ' chatmessage: ' + c)
           
            /* tests
             global["config"]["rcon"][this.tokenId]["instance"].say("Test from rcon")
            this.req["subscribeData"] = 
            {
              rcon: 1,
              serverId: user.serverId, 
              authKey: user.authKey, 
              messages: "Test from rcon",
              time: new Date().toISOString()
            };
            */
          }

          // send stuff back to the rcon id of the user where we found the keyword
        });
        utils.onSubscribeResponse(this.req,this.res);
      }

      utils.rcon_log(c, this.tokenId);
    });
  }.bind(this);

  this.getPlayers = function () {
    return new Promise((res, rej) => {
      this.rconCommand('players').then(res).catch(rej);
    });
  };

  this.getAdmins = function () {
    return new Promise((res, rej) => {
      this.rconCommand('admins').then(res).catch(rej);
    });
  };

  this.getPlayersArray = function () {
    return new Promise((res, rej) => {
      this.getPlayers().then((data) => {
        let dataArray = [
          ...data.matchAll(/(\d+)\s+(\b\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}):(\d+\b)\s+(\d+)\s+([0-9a-fA-F]+)\(\w+\)\s([\S ]+)$/gim),
        ].map((e) => e.splice(1, e.length - 1));
        dataArray.map((e) => {
          let name = e[5];
          if (name.includes(' (Lobby)')) {
            e[5] = name.replace(' (Lobby)', '');
            e[6] = true;
          } else {
            e[6] = false;
          }
        });

        res(dataArray);
      });
    });
  };

  this.getAdminsArray = function () {
    return new Promise((res, rej) => {
      this.getAdmins().then((data) => {
        let dataArray = [
          ...data.matchAll(/(\d+)\s+(\b\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}):(\d+\b)$/gim),
        ].map((e) => e.splice(1, e.length - 1));
        res(dataArray);
      });
    });
  };

  this.getPlayerCount = function () {
    return new Promise((res, rej) => {
      this.getPlayers()
        .then((data) => {
          res(data.split('\n').pop().match(/(\d+)/gim)[0]);
        })
        .catch(rej);
    });
  };

  this.say = function (message, player = -1) {
    return new Promise((res, rej) => {
      this.rconCommand(`say ${player} ${message}`).then(res).catch(rej);
    });
  };


  this.getBans = function () {
    return new Promise((res, rej) => {
      this.rconCommand('bans').then(res).catch(rej);
    });
  };

  this.getBansArray = function () {
    return new Promise((res, rej) => {
      this.getBans().then((data) => {
        let guidBans = [...data.matchAll(/(\d+)\s+([0-9a-fA-F]+)\s([perm|\d]+)\s+([\S ]+)$/gim)].map((e) => e.splice(1, e.length - 1));
        let ipBans = [...data.matchAll(/(\d+)\s+([0-9\.]+)\s+([perm|\d]+)\s+([\S ]+)$/gim)].map((e) => e.splice(1, e.length - 1));

        res([...guidBans.map((e) => ['guid', ...e]), ...ipBans.map((e) => ['ip', ...e])]);
      });
    });
  };
  this.rconCommand = function (command) {
    return new Promise((res, rej) => {
      try {
        if (!this.be) {
          rej('There is currently no Connection to an RCON Server.');
        }

        this.be.sendCommand(command, res);
      } catch (e) {
        rej(e);
      }
    });
  };
}

module.exports = Arma3Rcon;
